src/index.ts

```diff
@@ -17,13 +17,15 @@ export type Options = {
   strength?: number
   path?: string
   mnemonic?: string | null
+  source?: "import" | "newSeed"
```

Comment:
Rather than call this `source`, let's open up to arbitrary caller metadata. Here's my thought:

```
export type Options<MetadataType extends { [key: string]: string }> = {
  ...
  metadata?: MetadataType
}
...
export type SerializedHDKeyring<MetadataType extends { [key: string]: string }> = {
  ...
  metadata?: MetadataType
}
...
export default class HDKeyring<MetadataType extends { [key: string]: string }> implements Keyring<SerializedHDKeyring<MetadataType>> {
...
  readonly metadata: MetadataType

  constructor(options: Options = {}) ...
```

For e.g. Tally, this would be `HDKeyring<{ source: "import" | "newSeed" }>`.

@mhluongo does this align with what you were thinking as well?

Reply:
I'm not a huge fan of doing this at all — keyrings already have an ID they can be mapped against, and the smaller we keep this library, the smaller the audit surface we have.

If there's a big benefit to putting it here rather than the extension, I do prefer the more generic metadata... but it brings with it the issue of how it should be serialized to preserve type data.

Seeing "import" and "new seed" as type literals here now is definitely a code smell.

Reply:
I dig it - I think that perhaps the best way forward here is to change the `keyringSources` map in https://github.com/tallycash/extension/pull/1001 to `keyringMetadata` with @Shadowfiend 's comments above in mind - I will go ahead and do that.